### PR TITLE
Optimized resetting achievement criteria by condition 

### PR DESCRIPTION
**Pull request by ghost**
**Target branch(es):** 3.3.5

- [x] 3.3.5

**Issues addressed:** References #14274

**Tests performed:** Compiles, simple ingame tests

Optimized resetting achievement criteria by condition and changed the functionality to properly reset all possible criteria with given condition

### CODE REVIEW

**Comment by tkrokli** [comment]
(commenting here since I can't tag the affected line in the files list)
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Achievements/AchievementMgr.h:409:9: fatal error: 
      unknown type name 'AchievementCriteriaListByMiscValue'; did you mean
      'AchievementCriteriaListByCondition'?
        AchievementCriteriaListByMiscValue m_AchievementCriteriasByMiscValue[ACHIEVEMENT_CRITERIA_TYPE_TOTAL];
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        AchievementCriteriaListByCondition

/home/travis/build/TrinityCore/TrinityCore/src/server/game/Achievements/AchievementMgr.h:39:66: note: 
      'AchievementCriteriaListByCondition' declared here
typedef std::unordered_map<uint32, AchievementCriteriaEntryList> AchievementCriteriaListByCondition;
                                                                 ^
```
Affected line: https://github.com/TrinityCore/TrinityCore/pull/18905/files#diff-4d37329b2bc08ba1b22414c039c198b3R409

it is caused by the following line: https://github.com/xinef1/TrinityCore/blob/3.3.5achicond/src/server/game/Achievements/AchievementMgr.h#L39 
(PR file list view https://github.com/TrinityCore/TrinityCore/pull/18905/files#diff-4d37329b2bc08ba1b22414c039c198b3R39)

Looks like you either need to do something about line 408 + 409 in your version of `AchievementMgr.h` or see what you can do about the 10 places where `m_AchievementCriteriasByMiscValue` still exists in your PR code.

File: src/server/game/Achievements/AchievementMgr.cpp [hunk 1]
```diff
@@ -2289,6 +2284,13 @@ void AchievementGlobalMgr::LoadAchievementCriteriaList()
             }
         }
 
+        for (uint32 conditionField = 0; conditionField < MAX_CRITERIA_REQUIREMENTS; ++conditionField)
+        {
+            if (criteria->AdditionalRequirements[0].Type != ACHIEVEMENT_CRITERIA_CONDITION_NONE)
```
Comment (Shauren) [inline src/server/game/Achievements/AchievementMgr.cpp]: thats strange, only checking [0] in the loop?

File: src/server/game/Achievements/AchievementMgr.cpp [hunk 2]
```diff
@@ -2289,6 +2284,13 @@ void AchievementGlobalMgr::LoadAchievementCriteriaList()
...
+                if (conditionField == 0 || criteria->AdditionalRequirements[conditionField].Type != criteria->AdditionalRequirements[conditionField-1].Type)
+                    m_AchievementCriteriasByCondition[criteria->AdditionalRequirements[conditionField].Type][criteria->AdditionalRequirements[conditionField].Asset].push_back(criteria);
```
Comment (Shauren) [inline src/server/game/Achievements/AchievementMgr.cpp]: Add a helpful ASSERT on additional requirement type, just like the one for criteria types above

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Achievements/AchievementMgr.cpp (hunk 1 above)
Comment (xinef1) [inline src/server/game/Achievements/AchievementMgr.cpp]: Obviously a typo

File: src/server/game/Achievements/AchievementMgr.cpp [hunk 3]
```diff
@@ -2289,6 +2284,18 @@ void AchievementGlobalMgr::LoadAchievementCriteriaList()
...
         }
 
+        for (uint32 conditionField = 0; conditionField < MAX_CRITERIA_REQUIREMENTS; ++conditionField)
+        {
+            if (criteria->AdditionalRequirements[conditionField].Type != ACHIEVEMENT_CRITERIA_CONDITION_NONE)
+            {
+                ASSERT(criteria->AdditionalRequirements[conditionField].Type < ACHIEVEMENT_CRITERIA_CONDITION_MAX, "ACHIEVEMENT_CRITERIA_CONDITION_MAX must be greater than or equal to %u but is currently equal to %u",
+                    criteria->AdditionalRequirements[conditionField].Type + 1, ACHIEVEMENT_CRITERIA_CONDITION_MAX);
+
+                if (conditionField == 0 || criteria->AdditionalRequirements[conditionField].Type != criteria->AdditionalRequirements[conditionField-1].Type)
```
Comment (Shauren) [inline src/server/game/Achievements/AchievementMgr.cpp]: Is there a case where criteria->AdditionalRequirements[0].Type == criteria->AdditionalRequirements[1].Type but criteria->AdditionalRequirements[0].Asset != criteria->AdditionalRequirements[1].Asset

Comment (xinef1) [inline src/server/game/Achievements/AchievementMgr.cpp]: No, but I may change the if to check for that (patchs > 3.3.5 may differ in this matter)

**Pull request closed without merging**